### DISCUSSION TITLE
This implementation for every prototype is very hard to understand in direct.

### DISCUSSION THREAD

**Discussion by zxh19890103** [Leaflet/Leaflet#7847] [General]
https://github.com/Leaflet/Leaflet/blob/e3719a317ab325ed091fbbc6bfd34640fe19e13e/src/core/Class.js#L70

Though a long-time thinking about this block of code, there is another way to reach the same goal, I think. Just providing one function as bellow would avoid creating individual instances for every prototype:

```
proto.callInitHooks = callInitHooks;
```

```js
function callInitHooks() {
  if (this._initHooksCalled) {
    return;
  }

  this._initHooksCalled = true;

  let proto = this;
  // eslint-disable-next-line no-cond-assign
  while ((proto = proto.__proto__)) {
    if (proto === Object.prototype) break;
    if (proto.hasOwnProperty('_initHooks')) {
      for (var i = 0, len = proto._initHooks.length; i < len; i++) {
        proto._initHooks[i].call(this);
      }
    }
  }
}
```

How do you think of this idea?

**Comment by mourner** [comment]
The problem is that `__proto__` is only supported starting IE11 and later, and Leaflet currently supports older versions. We could switch to a more optimized implementation in future major versions after dropping support for everything below IE11.

**Comment by johnd0e** [comment]
Object.getPrototypeOf() is supported starting with IE9
